llmfoundry/data/finetuning/tasks.py

```diff
@@ -70,9 +96,70 @@ def _is_empty_or_nonexistent(dirpath: str) -> bool:
     return not os.path.isdir(dirpath) or len(os.listdir(dirpath)) == 0
 
 
-def _tokenize_formatted_example(
-        example: Dict[str, Any],
-        tokenizer: PreTrainedTokenizerBase) -> Dict[str, List[int]]:
+def _slice_chat_formatted_example(
+        example: ChatFormattedDict,
+        tokenizer: PreTrainedTokenizerBase) -> Tuple[str, str]:
+    """Slices the chat example into a formatted prompt and response.
+
+    Args:
+        example (ChatFormattedDict): The chat example containing the messages.
+        tokenizer (PreTrainedTokenizerBase): The tokenizer to apply the chat template.
+
+    Returns:
+        Tuple[str, str]: The prompt and response as separate strings.
+
+    Raises:
+        ValueError: If the chat example has less than two messages or if the last message is not from the assistant.
+        KeyError: If a message does not have a role or content.
+    """
+    messages = example['messages']
+
+    if len(messages) < 2:
+        raise ValueError(
+            f'chat example must have at least two messages. {messages=}')
+    last_message = messages[-1]
+    if last_message['role'] != 'assistant':
+        raise ValueError(
+            f'last message must be from assistant. {last_message=}')
+    for message in messages:
+        if 'role' not in message or 'content' not in message:
+            raise KeyError(f'message must have role and content. {message=}')
+
+    full_conversation = tokenizer.apply_chat_template(messages, tokenize=False)
+    prompt = tokenizer.apply_chat_template(messages[:-1],
```

Comment:
what happens if we do this and the tokenizer does not have a chat template?

Can you investigate and add a catch for that if appropriate?

Reply:
Ideally we would just apply the default chat template in the `tiktoken.py` if the tokenizer doesn't have one specified I think.

Reply:
This falls back to the default chat template (see https://github.com/huggingface/transformers/blob/d02d006cf315cf91e3a470eb72b9a9a7d0ecaf90/src/transformers/tokenization_utils_base.py#L1783C1-L1801C10 and https://github.com/huggingface/transformers/blob/d02d006cf315cf91e3a470eb72b9a9a7d0ecaf90/src/transformers/tokenization_utils_base.py#L1733C1-L1738C1)